### PR TITLE
Implement connection and subscription resources for jira and slack.

### PR DESCRIPTION
**Pull request by trentdibacco**
> [!WARNING]
> Dependent on https://github.com/hashicorp/terraform-provider-hcp/pull/1113 being merged first which upgrade the dependency of hcp-sdk-go to v0.116.0 and fixes a resource related to that change.

### :hammer_and_wrench: Description
__Preview__ of Manage a Vault Radar Connections and Subscription via the HCP Terraform Provider.

This add two resources that are tightly related to each other. Integration connections and integration subscriptions.
This introduces two types of connections and subscriptions with plans for more to come later.

The resources introduced in this PR are:
* hcp_vault_radar_integration_jira_connection
* hcp_vault_radar_integration_jira_subscription
* hcp_vault_radar_integration_slack_connection
* hcp_vault_radar_integration_slack_subscription

> [!Note] 
> My apologies in advance for making a PR that introduces more than one resource, but connection and subscriptions don't make sense on their own, and the difference between jira and slack is rather small and I wanted to prove out a way implement these two types with a common implementation.

> [!Note]  
> This is a preview, and my intent is for this resource not to show up in the in the [online documentations](https://registry.terraform.io/providers/hashicorp/hcp/latest/docs).
> Therefore, I did not set the **subcategory** in the markup.

## New Resource Checklist
- [x] __Plugin Framework__ 
- [ ] __Minimal LOC__: **(No, but they are tightly related)**.
- [x] __Acceptance Tests__: 
- [x] __Documentation__: 
- [x] __Well-formed Code__: `go fmt`. 

### Schema
- [x] __Uses Globally Unique ID__: Uses GUID returned by server.
- [x] __Validates Fields Where Possible__: Minimal
- [x] __Supports Optional Project ID Input__: 

### CRUD Operations
- [x] __Uses Context-Aware CRUD Functions__: 
- [x] __Uses Context For API Calls__: 
- [x] __Uses DRY API Calls__: 
- [X] __Handles Existing Resource Prior To Create__: 
- [x] __Refreshes Attributes During Read__: 
- [x] __Handles Removed Resource During Read__: 
- [x] __Handles Failed State During Read__:
- [x] __Handles Nonexistent Resource Prior To Delete__:

### Documentation
- [x] __Includes Descriptions For Resource And Fields__: 
- [x] __Includes Example__: 
- [x] __Includes Generated Docs__: 

### :building_construction: Acceptance tests

> [!Note] 
> These acceptance tests works, but requires external environments not avail to CI/CD.
> These tests are being skipped because of pre-existing conditions that would be require for the test to succeed. 
> - Requires Project already setup with Radar.
> - Requires a Service Account with an Admin role on the Project.
> - A Jira instance with access token
> - A Slack instance with bot token
> - Requires the following environment variables to be set:
>    - HCP_PROJECT_ID
>    - RADAR_INTEGRATION_JIRA_BASE_URL
>    - RADAR_INTEGRATION_JIRA_EMAIL
>    - RADAR_INTEGRATION_JIRA_TOKEN
>    - RADAR_INTEGRATION_JIRA_PROJECT_KEY
>    - RADAR_INTEGRATION_JIRA_ISSUE_TYPE
>    - RADAR_INTEGRATION_JIRA_ASSIGNEE
>    - RADAR_INTEGRATION_SLACK_TOKEN
>    - RADAR_INTEGRATION_SLACK_CHANNEL

> [!NOTE]
> I've created a ticket to note how we can address this in the future.
> https://hashicorp.atlassian.net/browse/VAULT-31051

- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch? 

Output from acceptance testing:

```sh
make testacc TESTARGS='-run=TestRadarIntegrationJiraConnection' 
...
=== RUN   TestRadarIntegrationJiraConnection
--- PASS: TestRadarIntegrationJiraConnection (14.06s)
...
```
```sh
make testacc TESTARGS='-run=TestRadarIntegrationJiraSubscription'
=== RUN   TestRadarIntegrationJiraSubscription
--- PASS: TestRadarIntegrationJiraSubscription (8.03s)
...
```
```sh
make testacc TESTARGS='-run=TestRadarIntegrationSlackConnection'
=== RUN   TestRadarIntegrationSlackConnection
--- PASS: TestRadarIntegrationSlackConnection (5.93s)
...
```
```sh
make testacc TESTARGS='-run=TestRadarIntegrationSlackSubscription'
=== RUN   TestRadarIntegrationSlackSubscription
--- PASS: TestRadarIntegrationSlackSubscription (7.73s)
...
```

### CODE REVIEW

File: internal/provider/vaultradar/resource_radar_integration_slack_connection.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+			Required:    true,
+			PlanModifiers: []planmodifier.String{
+				stringplanmodifier.RequiresReplace(),
+			},
+			Validators: []validator.String{
+				stringvalidator.LengthAtLeast(1),
+			},
+		},
+		"token": schema.StringAttribute{
+			Description: "Slack bot user oAuth token. Example: Bot token strings begin with 'xoxb'.",
```
Comment (nehalagrawal-hc) [inline internal/provider/vaultradar/resource_radar_integration_slack_connection.go:54]: nit OAuth*

**Review by kathytong-hashicorp — APPROVED** [review]

File: internal/provider/vaultradar/resource_radar_integration_jira_connection.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+			Validators: []validator.String{
+				stringvalidator.LengthAtLeast(1),
+			},
+		},
+		"email": schema.StringAttribute{
+			Description: `Jira user's email.`,
+			Required:    true,
+			PlanModifiers: []planmodifier.String{
+				stringplanmodifier.RequiresReplace(),
+			},
```
[the remark this answers is not in the document]
Comment (trentdibacco) [reply internal/provider/vaultradar/resource_radar_integration_jira_connection.go:59]: I don't see anyone else marking this a sensitive, but I could do that here.

[the remark this answers is not in the document]
Comment (kathytong-hashicorp) [reply internal/provider/vaultradar/resource_radar_integration_jira_connection.go:59]: If no one else is making emails sensitive we can leave it off, was just wondering 😅

File: internal/provider/vaultradar/integration_connection.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+		Details:         details,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError(errSummary, err.Error())
+		return
+	}
+
+	plan.SetID(types.StringValue(res.GetPayload().ID))
+	plan.SetProjectID(types.StringValue(projectID))
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
```
Comment (kartheek-hc) [inline internal/provider/vaultradar/integration_connection.go:128]: nit: Can we add trace logs similar to the Radar resource?
https://github.com/hashicorp/terraform-provider-hcp/blob/cd7cff2c530a9c8d65d26f41ebf659a23501e056/internal/provider/vaultradar/resource_radar_source_github_enterprise.go#L149

File: examples/resources/hcp_vault_radar_integration_jira_subscription/resource.tf [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  token    = var.jira_token
+  base_url = "https://example.atlassian.net"
+}
+
+resource "hcp_vault_radar_integration_jira_subscription" "jira_subscription" {
+  name             = "example integration jira subscription"
+  connection_id    = hcp_vault_radar_integration_jira_connection.jira_connection.id
+  jira_project_key = "SEC"
+  issue_type       = "Task"
+  assignee         = "id-of-assignee"
```
[the remark this answers is not in the document]
Comment (trentdibacco) [reply examples/resources/hcp_vault_radar_integration_jira_subscription/resource.tf:19]: I'll match the example I gave in the description.

File: internal/provider/vaultradar/resource_radar_integration_jira_connection.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (trentdibacco) [reply internal/provider/vaultradar/resource_radar_integration_jira_connection.go:59]: As its part of the auth, I think I agree with you. Marking it as sensitive just keeps it from printing in any logs.

File: docs/resources/vault_radar_integration_jira_connection.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  type      = string
+  sensitive = true
+}
+
+resource "hcp_vault_radar_integration_jira_connection" "jira_connection" {
+  name     = "example connection to jira"
+  email    = "jane.smith@example.com"
+  token    = var.jira_token
+  base_url = "https://example.atlassian.net"
+}
```
[the remark this answers is not in the document]
Comment (trentdibacco) [reply docs/resources/vault_radar_integration_jira_connection.md:27]: Not sure what you mean by jira server, and I went by what was required in the integration jira connection in the UI and the python schema.

File: docs/resources/vault_radar_integration_jira_connection.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
+---
+page_title: "hcp_vault_radar_integration_jira_connection Resource - terraform-provider-hcp"
+subcategory: ""
+description: |-
+  This terraform resource manages an Integration Jira Connection in Vault Radar.
+---
+
+# hcp_vault_radar_integration_jira_connection (Resource)
```
[the remark this answers is not in the document]
Comment (trentdibacco) [reply docs/resources/vault_radar_integration_jira_connection.md:8]: I choose integration to differentiate source related stuff.
Example: 
For sources I've using  hcp_vault_radar_source_XXX like hcp_vault_radar_source_github_cloud

File: docs/resources/vault_radar_integration_jira_subscription.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+```
+
+
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- `connection_id` (String) id of the integration jira connection to use for the subscription.
+- `issue_type` (String) The type of issue to be created from the alert(s). Example: Task
```
[the remark this answers is not in the document]
Comment (trentdibacco) [reply docs/resources/vault_radar_integration_jira_subscription.md:47]: I copy the text from our UI. But I can update it to events instead.

File: docs/resources/vault_radar_integration_slack_subscription.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+}
+```
+
+
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- `channel` (String) Slack channel that messages will be sent to. Note that HashiCorp Vault Radar will send a test message to verify channel's name. Example: dev-ops-team
```
[the remark this answers is not in the document]
Comment (trentdibacco) [reply docs/resources/vault_radar_integration_slack_subscription.md:41]: Again, I just copied what the UI is showing today.
But I like what you have, will use that.

File: docs/resources/vault_radar_integration_jira_connection.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kartheek-hc) [reply docs/resources/vault_radar_integration_jira_connection.md:27]: Jira server is the self-hosted variant of Atlassian Jira. I am not sure we currently support Jira server but going by the integration type (`jira`), I thought we want to use the same terraform resource for both Jira Cloud and Server. If not, the resource name should contain `jira_cloud` right

File: internal/provider/vaultradar/integration_connection.go (hunk 3 above)
Comment (trentdibacco) [reply internal/provider/vaultradar/integration_connection.go:128]: Actually I was planning on removing those. I had initially put them there just so I could confirm what was going on, but there is enough in the logs that make these log unnecessary for a success state.

File: docs/resources/vault_radar_integration_jira_subscription.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (kartheek-hc) [reply docs/resources/vault_radar_integration_jira_subscription.md:47]: Oh ok, I think we need to update the UI as well and be consistent

**Review by kartheek-hc — APPROVED** [review]

File: docs/resources/vault_radar_integration_jira_connection.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kathytong-hashicorp) [reply docs/resources/vault_radar_integration_jira_connection.md:27]: I think Jira also [supports Basic Auth](https://developer.atlassian.com/server/jira/platform/basic-authentication/), we may have never made the update to support personal access tokens?

[the remark this answers is not in the document]
Comment (trentdibacco) [reply docs/resources/vault_radar_integration_jira_connection.md:27]: K, going to leave it as like this then.

[the remark this answers is not in the document]
Comment (kartheek-hc) [reply docs/resources/vault_radar_integration_jira_connection.md:27]: Basic auth is not preferred way to make requests to the Jira server API. Same story for Confluence server onboarding, we only support personal access token not basic auth

> However, as basic authentication repeatedly sends the username and password on each request, which could be cached in the web browser, it is not the most secure method of authentication we support.

[the remark this answers is not in the document]
Comment (kartheek-hc) [reply docs/resources/vault_radar_integration_jira_connection.md:27]: > I think Jira Server also [supports Basic Auth](https://developer.atlassian.com/server/jira/platform/basic-authentication/), we may have never made the update to support personal access tokens?

I think the APIs that we use validate the assignee field on the subscription only work for Jira Cloud. We probably have to make some changes before Jira server integration is fully supported

**Review by jasonpilz — APPROVED** [review]
LGTM

**Pull request merged**